### PR TITLE
Enable support of LeanElasticPeakWorkspace in SortPeakWorkspace and CompareWorkspaces

### LINKED ISSUE #30974: Fix table sort in LeanElasticPeaksWorkspace/SortPeaksWorkspace and CompareWorkspaces
Sorting and comparison of LeanElasticPeakWorkspaces are not functioning

```
Python-[Error] Traceback (most recent call last):
  File "/home/rwp/src/mantid/qt/python/mantidqt/widgets/workspacedisplay/table/presenter.py", line 268, in action_sort
    self.presenter.model.sort(selected_column, sort_ascending)
  File "/home/rwp/src/mantid/qt/python/mantidqt/widgets/workspacedisplay/table/model.py", line 156, in sort
    SortPeaksWorkspace(InputWorkspace=self.ws,
  File "/home/rwp/src/mantid/Framework/PythonInterface/mantid/simpleapi.py", line 1034, in __call__
    set_properties(algm, *args, **final_keywords)
  File "/home/rwp/src/mantid/Framework/PythonInterface/mantid/simpleapi.py", line 952, in set_properties
    do_set_property(key, value)
  File "/home/rwp/src/mantid/Framework/PythonInterface/mantid/simpleapi.py", line 933, in do_set_property
    alg_object.setPropertyValue(key, new_value.name())
ValueError: Invalid value for property InputWorkspace (PeaksWorkspace) from string "peaks": Workspace peaks is not of the correct type
```

The two algorithm, 
[SortPeaksWorkspace](https://docs.mantidproject.org/nightly/algorithms/SortPeaksWorkspace-v1.html) and 
[CompareWorkspaces](https://docs.mantidproject.org/nightly/algorithms/CompareWorkspaces-v1.html), need to be adjusted to ensure that they also support LeanElasticPeakWorkspaces.

### PR DESCRIPTION
**Pull request by KedoKudo**
**Description of work.**

This PR enables support of LeanElasticPeakWorkspace for both `SortPeakWorkspace` and `CompareWorkspaces`.

- Related Gitlab task: [263](https://code.ornl.gov/sns-hfir-scse/diffraction/single-crystal/single-crystal-diffraction/-/issues/263)
- This is #30973 into `master`

**To test:**

Use the following for testing

```
from mantid.simpleapi import *
from mantid.geometry import OrientedLattice
from itertools import permutations
import numpy as np

ol=OrientedLattice(5,7,12,90,90,120)
ub=ol.getUB()
print(ub)

peaks = CreatePeaksWorkspace(NumberOfPeaks=0)

for hkl in set(permutations([1,1,1,0,0,0,-1,-1,-1],3)):
    p = peaks.createPeakQSample(2*np.pi*np.dot(ub,hkl))
    peaks.addPeak(p)

sorted_peaks = SortPeaksWorkspace(peaks, ColumnNameToSortBy='DSpacing')

CompareWorkspaces("peaks", "sorted_peaks")
```

Fixes #30974 . 

*This does not require release notes* because **LeanElasticPeakWorkspace related topic will be addressed in a separate PR once LeanEalsticPeak is fully implemented.**

---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

### CODE REVIEW

File: Framework/Algorithms/src/CompareWorkspaces.cpp [hunk 1]
```diff
@@ -1134,6 +1162,91 @@ void CompareWorkspaces::doPeaksComparison(PeaksWorkspace_sptr tws1,
...
+      } else if (name == "l") {
+        s1 = ipws1->getPeak(i).getL();
+        s2 = ipws2->getPeak(i).getL();
+      } else if (name == "wavelength") {
+        s1 = ipws1->getPeak(i).getWavelength();
+        s2 = ipws2->getPeak(i).getWavelength();
+      } else if (name == "energy") {
+        s1 = ipws1->getPeak(i).getFinalEnergy();
+        s2 = ipws2->getPeak(i).getFinalEnergy();
+      } else if (name == "tof") {
```
Comment (rosswhitfield) [inline Framework/Algorithms/src/CompareWorkspaces.cpp:1218]: `tof`, `row`, and `cow` aren't columns in the LeanElasticPeaksWorkspace.

I think you should also somehow check the `QSample` column because otherwise, I could have two different peaks with the same spacing but different q vectors that would still match.

**Review by rosswhitfield — CHANGES_REQUESTED** [review]

File: Framework/Algorithms/src/CompareWorkspaces.cpp (hunk 1 above)
Comment (KedoKudo) [reply Framework/Algorithms/src/CompareWorkspaces.cpp:1218]: good point, I will adjust the checking now

File: Framework/Algorithms/src/CompareWorkspaces.cpp [hunk 2]
```diff
@@ -1081,7 +1109,7 @@ void CompareWorkspaces::doPeaksComparison(PeaksWorkspace_sptr tws1,
       std::string name = col->name();
       double s1 = 0.0;
       double s2 = 0.0;
-      if (name == "runnumber") {
+      if (name == "RunNumber") {
```
Comment (rosswhitfield) [inline Framework/Algorithms/src/CompareWorkspaces.cpp:1112]: So this didn't work correctly before for PeaksWorkspace :open_mouth: 

A quick test show you did fix this for PeaksWorkspace, this show matched on master but mismatch on this branch :+1: 
```python
ws = CreateSampleWorkspace()
peaks = CreatePeaksWorkspace(InstrumentWorkspace='ws')
peaks2 = CreatePeaksWorkspace(InstrumentWorkspace='ws')
peaks.getPeak(0).setRunNumber(123)
CompareWorkspaces(peaks, peaks2)
```

It seems some of the other column names are also incorrect for PeakWorkspace

`detid` -> `DetID`
`tof` -> `TOF`
`row` -> `Row`
`col` -> `Col`

So I'm guessing this means there is no real testing for CompareWorkspaces with PeakWorkspace :facepalm:

File: Framework/Algorithms/src/CompareWorkspaces.cpp [hunk 3]
```diff
@@ -1134,6 +1162,90 @@ void CompareWorkspaces::doPeaksComparison(PeaksWorkspace_sptr tws1,
...
+      } else if (name == "k") {
+        s1 = ipws1->getPeak(i).getK();
+        s2 = ipws2->getPeak(i).getK();
+      } else if (name == "l") {
+        s1 = ipws1->getPeak(i).getL();
+        s2 = ipws2->getPeak(i).getL();
+      } else if (name == "Wavelength") {
+        s1 = ipws1->getPeak(i).getWavelength();
+        s2 = ipws2->getPeak(i).getWavelength();
+      } else if (name == "Energy") {
```
Comment (rosswhitfield) [inline Framework/Algorithms/src/CompareWorkspaces.cpp:1215]: We don't need to check the `Energy` column because in LeanElasticPeak it is just calculated directly from `Wavelength`.

File: Framework/Algorithms/src/CompareWorkspaces.cpp [hunk 4]
```diff
@@ -1134,6 +1162,90 @@ void CompareWorkspaces::doPeaksComparison(PeaksWorkspace_sptr tws1,
...
+      } else if (name == "Intens") {
+        s1 = ipws1->getPeak(i).getIntensity();
+        s2 = ipws2->getPeak(i).getIntensity();
+      } else if (name == "SigInt") {
+        s1 = ipws1->getPeak(i).getSigmaIntensity();
+        s2 = ipws2->getPeak(i).getSigmaIntensity();
+      } else if (name == "BinCount") {
+        s1 = ipws1->getPeak(i).getBinCount();
+        s2 = ipws2->getPeak(i).getBinCount();
+      } else if (name == "QLab") {
```
Comment (rosswhitfield) [inline Framework/Algorithms/src/CompareWorkspaces.cpp:1230]: Can we also check `QSample`, there are cases where, for example with four-circle, the `QLab` is almost identical for every peak but the `QSample` is different because the goniometer is different.

**Review by rosswhitfield — CHANGES_REQUESTED** [review]
SortPeaksWorkspace works, just some changes for CompareWorkspaces.

**Comment by rosswhitfield** [comment]
Also by fixing CompareWorksapce for the PeaksWorkspace a lot of system tests are now failing :joy:

File: Framework/Algorithms/src/CompareWorkspaces.cpp (hunk 2 above)
Comment (KedoKudo) [reply Framework/Algorithms/src/CompareWorkspaces.cpp:1112]: That seems to be the case. Also, I have updated the column names as suggested.

File: Framework/Algorithms/src/CompareWorkspaces.cpp (hunk 3 above)
Comment (KedoKudo) [reply Framework/Algorithms/src/CompareWorkspaces.cpp:1215]: Removed `energy` from checklist.

File: Framework/Algorithms/src/CompareWorkspaces.cpp (hunk 4 above)
Comment (KedoKudo) [reply Framework/Algorithms/src/CompareWorkspaces.cpp:1230]: Added `QSample` to checklist as suggested.

**Comment by KedoKudo** [comment]
> Also by fixing CompareWorksapce for the PeaksWorkspace a lot of system tests are now failing

The default tolerance for CompareWorkspace is 0, so any float type entry comparison will leads to an error, for example

```
CompareWorkspaces-[Notice] CompareWorkspaces started
CompareWorkspaces-[Notice] Wavelength
CompareWorkspaces-[Notice] s1 = 0.535548
CompareWorkspaces-[Notice] s2 = 0.535548
CompareWorkspaces-[Notice] std::fabs(s1 - s2) = 1.11022e-16
CompareWorkspaces-[Notice] tolerance = 0
CompareWorkspaces-[Notice] Data mismatch at cell (row#,col#): (4,5)
CompareWorkspaces-[Notice] The workspaces did not match: Data mismatch
```

I am changing the default tolerance to 1e-15, and hopefully this will help the system tests that forget to specify tolerance to pass.

**Comment by KedoKudo** [comment]
> > Also by fixing CompareWorksapce for the PeaksWorkspace a lot of system tests are now failing
> 
> The default tolerance for CompareWorkspace is 0, so any float type entry comparison will leads to an error, for example
> 
> ```
> CompareWorkspaces-[Notice] CompareWorkspaces started
> CompareWorkspaces-[Notice] Wavelength
> CompareWorkspaces-[Notice] s1 = 0.535548
> CompareWorkspaces-[Notice] s2 = 0.535548
> CompareWorkspaces-[Notice] std::fabs(s1 - s2) = 1.11022e-16
> CompareWorkspaces-[Notice] tolerance = 0
> CompareWorkspaces-[Notice] Data mismatch at cell (row#,col#): (4,5)
> CompareWorkspaces-[Notice] The workspaces did not match: Data mismatch
> ```
> 
> I am changing the default tolerance to 1e-15, and hopefully this will help the system tests that forget to specify tolerance to pass.

Unfortunately there are other systems failing due to actual data mismatch
- FindSatellitePeaksTest.FindSatellitePeaksTestFixedNumQ
```
CompareWorkspaces-[Notice] CompareWorkspaces started
CompareWorkspaces-[Information] Algorithm: CompareWorkspaces v1
CompareWorkspaces-[Information] Execution Date: 2021-03-26 20:33:28.084038000
CompareWorkspaces-[Information] Execution Duration: -1 seconds
CompareWorkspaces-[Information] UUID: c98007a0-40f5-40d1-a422-523f262a5a12
CompareWorkspaces-[Information] Parameters:
CompareWorkspaces-[Information]   Name: Workspace1, Value: refine_peaks_test, Default?: No, Direction: Input
CompareWorkspaces-[Information]   Name: Workspace2, Value: RefFile, Default?: No, Direction: Input
CompareWorkspaces-[Information]   Name: Tolerance, Value: 1e-08, Default?: No, Direction: Input
CompareWorkspaces-[Information]   Name: CheckType, Value: 1, Default?: Yes, Direction: Input
CompareWorkspaces-[Information]   Name: CheckAxes, Value: 1, Default?: Yes, Direction: Input
CompareWorkspaces-[Information]   Name: CheckSpectraMap, Value: 1, Default?: Yes, Direction: Input
CompareWorkspaces-[Information]   Name: CheckInstrument, Value: 1, Default?: Yes, Direction: Input
CompareWorkspaces-[Information]   Name: CheckMasking, Value: 1, Default?: Yes, Direction: Input
CompareWorkspaces-[Information]   Name: CheckSample, Value: 0, Default?: Yes, Direction: Input
CompareWorkspaces-[Information]   Name: ToleranceRelErr, Value: 0, Default?: Yes, Direction: Input
CompareWorkspaces-[Information]   Name: CheckAllData, Value: 0, Default?: Yes, Direction: Input
CompareWorkspaces-[Information]   Name: NumberMismatchedSpectraToPrint, Value: 1, Default?: Yes, Direction: Input
CompareWorkspaces-[Information]   Name: DetailedPrintIndex, Value: 2147483647, Default?: Yes, Direction: Input
CompareWorkspaces-[Information]   Name: Result, Value: 0, Default?: Yes, Direction: Output
CompareWorkspaces-[Information]   Name: Messages, Value: compare_msgs, Default?: Yes, Direction: Output
CompareWorkspaces-[Notice] BinCount
CompareWorkspaces-[Notice] s1 = 60818
CompareWorkspaces-[Notice] s2 = 0
CompareWorkspaces-[Notice] std::fabs(s1 - s2) = 60818
CompareWorkspaces-[Notice] tolerance = 1e-08
CompareWorkspaces-[Notice] Data mismatch at cell (row#,col#): (0,12)
CompareWorkspaces-[Notice] The workspaces did not match: Data mismatch
CompareWorkspaces-[Notice] CompareWorkspaces successful, Duration 0.00 seconds
```

File: Framework/Algorithms/src/CompareWorkspaces.cpp [hunk 5]
```diff
@@ -136,7 +137,7 @@ void CompareWorkspaces::init() {
                   "The name of the second input workspace.");
 
   declareProperty(
-      "Tolerance", 0.0,
+      "Tolerance", 1e-10,
```
Comment (rosswhitfield) [inline Framework/Algorithms/src/CompareWorkspaces.cpp:140]: Can you revert this change as well since it's not needed for this PR.

**Review by rosswhitfield — CHANGES_REQUESTED** [review]

File: Framework/Algorithms/src/CompareWorkspaces.cpp (hunk 5 above)
Comment (KedoKudo) [reply Framework/Algorithms/src/CompareWorkspaces.cpp:140]: Got it, I will do it now.

**Review by rosswhitfield — DISMISSED** [review]

**Review by rosswhitfield — APPROVED** [review]
